tower-balance/src/p2c/service.rs

```diff
@@ -255,77 +165,39 @@ where
 
     /// Prepares the balancer to process a request.
     ///
-    /// When `Async::Ready` is returned, `ready_index` is set with a valid index
-    /// into `ready` referring to a `Service` that is ready to disptach a request.
+    /// When `Async::Ready` is returned, `next_ready_index` is set with a valid
+    /// index into `ready` referring to a `Service` that is ready to disptach a
+    /// request.
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
         // First and foremost, process discovery updates. This removes or updates a
         // previously-selected `ready_index` if appropriate.
         self.poll_discover()?;
 
         // Drive new or busy services to readiness.
         self.poll_unready();
-        trace!({ nready = self.ready_services.len(), nunready = self.unready_services.len() }, "poll_ready");
 
+        let mut index = self.services.take_next_ready_index();
         loop {
-            // If a node has already been selected, ensure that it is ready.
+            // If a service has already been selected, ensure that it is ready.
             // This ensures that the underlying service is ready immediately
             // before a request is dispatched to it. If, e.g., a failure
             // detector has changed the state of the service, it may be evicted
-            // from the ready set so that P2C can be performed again.
-            if let Some(index) = self.next_ready_index {
-                trace!({ next.idx = index }, "preselected ready_index");
-                debug_assert!(index < self.ready_services.len());
-
-                if let Ok(Async::Ready(())) = self.poll_ready_index_or_evict(index) {
+            // from the ready set so that another service can be selected.
+            if let Some(index) = index {
+                if let Ok(Async::Ready(())) = self.services.poll_next_ready_index(index) {
```

Comment:
I assume this sets `next_ready_index` internally in `self.services` again given that we "take" it above?

Reply:
Yeah. I'm updating the docs for poll_next_ready_index to be:

```rust
    /// Configures the ready service to be used by `call`.
    ///
    /// This may be called instead of `Service::poll_ready`.
    ///
    /// If the service is no longer ready and there are no equivalent unready
    /// services, it is moved back into the ready set and `Async::NotReady` is
    /// returned.
    ///
    /// If the service errors, it is removed and dropped and the error is returned.
    ///
    /// Otherwise, the `next_ready_index` is set with the provided index and
    /// `Async::Ready` is returned.
```

If you have suggestions on improving this, please have at it.

Reply:
What does "equivalent unready service" mean here?

Also, is the concept of `next_ready_index` documented anywhere?